### PR TITLE
hw10 done

### PR DESCRIPTION
**Pull request by emineyvazov222**

### CODE REVIEW

File: src/main/java/org/spring/FullVersion/CollectionFamilyDao.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package org.spring.FullVersion;
 
 
+import java.io.*;
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/CollectionFamilyDao.java:4]: avoid wildcard import

File: src/main/java/org/spring/FullVersion/CollectionFamilyDao.java [hunk 2]
```diff
@@ -48,6 +49,27 @@ public void saveFamily(Family family) {
 
     }
 
+    @Override
+    public void saveData(List<Family> families) {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream("families.dat"))) {
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/CollectionFamilyDao.java]: You can create a variable FILE_PATH, and set it families.dat, and use it instead write it every time

File: src/main/java/org/spring/FullVersion/CollectionFamilyDao.java [hunk 3]
```diff
@@ -48,6 +49,27 @@ public void saveFamily(Family family) {
...
+        }
+
+    }
+
+    @Override
+    public List<Family> loadData() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream("families.dat"))) {
+            families = (List<Family>) ois.readObject();
+            System.out.println("Data loaded from file.");
+        } catch (IOException | ClassNotFoundException e) {
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/CollectionFamilyDao.java]: you can handle this exceptions in service layer

File: src/main/java/org/spring/FullVersion/Logger.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package org.spring.FullVersion;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class Logger {
+    private static final String LOG_FILE = "application.log";
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/Logger.java:12]: pld leave a space

File: src/main/java/org/spring/FullVersion/Logger.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+
+    public static void error(String message) {
+        log("[ERROR]", message);
+    }
+
+
+    private static void log(String logType, String message) {
+        String formattedMessage = String.format("%s %s %s", dateFormat.format(new Date()), logType, message);
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/Logger.java]: use local date instead date

File: src/main/java/org/spring/FullVersion/Logger.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    private static final SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy HH:mm");
+
+
+    public static void info(String message) {
+        log("[DEBUG]", message);
+    }
+
+
+    public static void error(String message) {
+        log("[ERROR]", message);
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/Logger.java]: you can give these [] on directly in formatted message like [%s]

**Review by sabuhihuseynov — CHANGES_REQUESTED** [review]
pls look

File: src/main/java/org/spring/FullVersion/FamilyService.java [hunk 7]
```diff
@@ -124,5 +125,19 @@ public void addPet(int index, Pet pet) {
...
+
+    public void loadFamilies() {
+        List<Family> loadedFamilies = familyDao.loadData();
+        if (loadedFamilies != null) {
+            familyDao.saveData(loadedFamilies);
+            Logger.info("Families successfully loaded from file.");
+        }
+
+    }
+}
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/FamilyService.java:153]: you can catch possible exceptions here

File: src/main/java/org/spring/FullVersion/FamilyOverflowException.java [hunk 8]
```diff
@@ -1,7 +1,7 @@
 package org.spring.FullVersion;
 
-public class FamilyOverflowException extends RuntimeException {
-    public FamilyOverflowException(String message) {
+public class FamilyOverflowException extends Exception {
```
Comment (sabuhihuseynov) [inline src/main/java/org/spring/FullVersion/FamilyOverflowException.java]: why extends Exception instead RuntimeException?

**Review by sabuhihuseynov — CHANGES_REQUESTED** [review]
pls look

File: src/main/java/org/spring/FullVersion/FamilyService.java (hunk 7 above)
Comment (emineyvazov222) [reply src/main/java/org/spring/FullVersion/FamilyService.java:153]: I don't know how to do it
@sabuhihuseynov

**Pull request merged**